### ISSUE TITLE
Unicode Case Folding

### ISSUE THREAD

**Issue by iSazonov** [dotnet/corefxlab#2610]
Come from https://github.com/dotnet/corefx/issues/17233
Related discussion https://github.com/PowerShell/PowerShell/issues/8120
~~First alfa code https://github.com/iSazonov/PowerShell/tree/add-unicode3/src/System.Management.Automation/utils/unicode~~
Beta code and help utils https://github.com/iSazonov/SCF

Initial PR #2637.

# Suggestion

S1. Implement internal API for case-insensitive language-neitral comparisions and convertions based on Unicode Simple case folding.
S2. Expose public API like `System.StringComparison.SimpleCaseFolding` for `String.Compare()` and `String.IndexOf()`. And similar public API for Span<char>.
Update: Improve OrdinalIgnoreCase comparisons everywhere. 
S3. Expose public API for `System.String` to convert to simple case folding format. And similar public API for Span<char>.

## Additional considerations

Initially all suggestions is only for Utf16. Utf8 (and others) could be considered too.

Initially all suggestions is only for Simple Case Folding. Full Case Folding could be useful for culture-sensetive comparisions (specially for Regex).

# Expected benefits

Increase performance of case-insensitive language-neitral comparisions for strings and Span<char>. 

Get benefits in follow core APIs:
- Regex case-insensitive search (invariant culture option)
- string case-insensitive language-neitral comparisions (speed up OrdinalIgnoreCase comparisons everywhere)
- Span<char> buffer case-insensitive language-neitral comparisions
- Speed up Marvin.ComputeHash32OrdinalIgnoreCase()

Get benefits for case-insensetive languages like PowerShell, HTTP, HTML, CSS, URL, and others, based on Unicode standard definition of _Equivalent Case-Insensitive Identifiers_.

# Performance measures

First results from a [PowerShell repo prototype](https://github.com/PowerShell/PowerShell/issues/8120#issuecomment-435045448) showed that the simple case folding comparision performance is comparable to String.Ordinal comparision performance.

# Justification

Unicode standard introduces the concept of _Simple case folding_ that is intended to compare strings in case-insensetive and language-neitral manner.

## Short quotes from Unicode standard

From: http://unicode.org/faq/casemap_charprop.html

> Q: What is the difference between case mapping and case folding?
A: Case mapping or case conversion is a process whereby strings are converted to a particular form—uppercase, lowercase, or titlecase—possibly for display to the user. Case folding is mostly used for caseless comparison of text, such as identifiers in a computer program, rather than actual text transformation. Case folding in Unicode is primarily based on the lowercase mapping, but includes additional changes to the source text to help make it language-insensitive and consistent. As a result, case-folded text should be used solely for internal processing and generally should not be stored or displayed to the end user.

From http://www.unicode.org/reports/tr31/tr31-29.html#R5

> Equivalent Case-Insensitive Identifiers: To meet this requirement, an implementation shall specify either simple or full case folding, and adhere to the Unicode specification for that folding. Any two identifiers that have the same case-folded form shall be treated as equivalent by the implementation. 

From http://www.unicode.org/reports/tr21/tr21-5.html

> 1.3 Caseless Matching
Caseless matching is implemented using case-folding. The latter is the process of mapping strings to a canonical form where case differences are erased. Case-folding allows for fast caseless matches in lookups, since only binary comparison is required. Case-folding is more than just conversion to lowercase. For example, it handles cases such as the Greek sigma, so that  "Μάϊος" and "ΜΆΪΟΣ" will match correctly.

From http://unicode.org/reports/tr18/#Folded_Matching

> 3.10 Folded Matching (Retracted) 
RL3.10 Folded Matching 
Previous versions of RL3.10 described tailored folding. However, for most full-featured regular expression engines, it is quite difficult to match under folding equivalences that are not 1:1. For more discussion of this, see 1.5 Simple Loose Matches and 2.1 Canonical Equivalents. Thus RL3.10 has been retracted. 

From http://unicode.org/reports/tr18/#Simple_Loose_Matches

> 1.5  Simple Loose Matches 
Most regular expression engines offer caseless matching as the only loose matching. If the engine does offers this, then it needs to account for the large range of cased Unicode characters outside of ASCII.
RL1.5 Simple Loose Matches 
 To meet this requirement, if an implementation provides for case-insensitive matching, then it shall provide at least the simple, default Unicode case-insensitive matching, and specify which properties are closed and which are not. 
To meet this requirement, if an implementation provides for case conversions, then it shall provide at least the simple, default Unicode case folding. 

From http://www.unicode.org/versions/Unicode11.0.0/ch03.pdf

> 3.13 Default Case Algorithms

>Default Case Folding
Default Case Folding is based on the full case conversion operations without the contextdependent
mappings sensitive to the casing context.
R4 toCasefold(X): Map each character C in X to Case_Folding(C).
• Case_Folding(C) uses the mappings with the status field value “C” or “F” in the
data file CaseFolding.txt in the Unicode Character Database.

A modified form of Default Case Folding is designed for best behavior when doing caseless
matching of strings interpreted as identifiers.
(See above http://www.unicode.org/reports/tr31/tr31-29.html#R5)

>Default Caseless Matching
Default caseless matching is the process of comparing two strings for case-insensitive
equality. The definitions of Unicode Default Caseless Matching build on the definitions of
Unicode Default Case Folding.
Default Caseless Matching uses full case folding:
D144 A string X is a caseless match for a string Y if and only if:
toCasefold(X) = toCasefold(Y)

/cc @tarekgh @ahsonkhan @KrzysztofCwalina  (Sorry if my ping is wrong)

**Comment by ahsonkhan** [comment]
Based on https://github.com/dotnet/corefx/issues/17233#issuecomment-434909772, do you plan to produce this functionality in a stand-alone NuGet package? If so, do you have a package name/namespace name in mind?

**Comment by tarekgh** [comment]
You are including the whole BMP Unicode plane data which is too much. 

I think we can optimize storing the data in better way which will reduce the size. you can use the 8-4-4 mapping technique.

**Comment by iSazonov** [comment]
> do you plan to produce this functionality in a stand-alone NuGet package? If so, do you have a package name/namespace name in mind?

The best target for me is to get the API, at least `System.StringComparison.SimpleCaseFolding, in .Net Core 3.0. Is it possible implement `System.StringComparison.SimpleCaseFolding` in the repo? If yes the package name/namespace name would be obvious. If no I still can not imagine where it is better to place and how name. Perhaps `System.Globalization` is a good place for all helper methods and cache arrays and rest APIs would be extension methods for String and Span<char> types.

> you can use the 8-4-4 mapping technique.

Yes, I agree that we should reduce size the cache array. Could you please point me a sample code?

**Comment by tarekgh** [comment]
> Could you please point me a sample code?

I have crafted some sample show how to read the case folding Unicode data and then generate and access the size optimized tables https://gist.github.com/tarekgh/55dfaf0f44689738c3a6ca67941ccdc2

**Comment by iSazonov** [comment]
@tarekgh Thanks!
I calculated the first byte for both 1 and 2 panes (ignore first "1" ) and had 18+5=23 unique values. It is 23*256 = 5888 elements. Total size 5888+256=6153.
Could it be easier implement and faster?

```
Count Name
----- ----
   57 00
  131 01
   39 02
   66 03
  124 04
   62 05
   40 10
    6 13
   55 1C
  125 1E
   97 1F
   21 21
   26 24
  115 2C
   37 A6
   72 A7
   80 AB
   26 FF

Count Name
----- ----
   76 04
   51 0C
   32 18
   32 6E
   34 E9
```

**Comment by tarekgh** [comment]
@iSazonov could you explain more what you mean?

**Comment by tarekgh** [comment]
do you mean you make it 2 lookup tables instead of 3? this may work too if the size difference is small. should be simpler too.

**Comment by iSazonov** [comment]
From CaseFolding.txt  I take first byte and count unique values. Yes, it will be 2 levels. Perhaps in this case there will be less pointer transformations and memory accesses.

**Comment by tarekgh** [comment]
you can try and measure.

**Comment by iSazonov** [comment]
Will try.

In my alfa code I tried cache array of ushort type and catch slow type cast from ushort to char so I began to use arrays of char type. I see you use int[] in your gist. What is more correct way?

**Comment by tarekgh** [comment]
I'll not be able to help before the new year :-) as I am currently off. but you got the idea and it is easy to generate the needed data.

**Comment by iSazonov** [comment]
I created code with 2-level cache https://github.com/PowerShell/PowerShell/pull/8515 
but I just can’t get acceptable performance.

**Comment by iSazonov** [comment]
@tarekgh  I got some performance progress for SMP. Then I started testing the code for surrogates and discover that your gist code works only for SMP. Do you have any thoughts how best to handle them?

**Comment by tarekgh** [comment]
@iSazonov for surrogates we'll need to generate a tables for that too the same way. I didn't look but if surrogates don't have many folding cases we can special case those and handle then differently. we'll just need to look how many folding ranges there. if I get some time I can try to take a look.

**Comment by iSazonov** [comment]
Thanks @tarekgh !  There is ~225 codepoints from Pane1 in SimpleCase.txt. The stats I published above https://github.com/dotnet/corefxlab/issues/2610#issuecomment-443957930

**Comment by tarekgh** [comment]
looking at https://www.unicode.org/Public/5.2.0/ucd/CaseFolding.txt I am seeing it is just one range from 10400 ~ 10427 so we can have just a simple check like 

if (c is Surrogate and in range 0x10400 ~ 0x10427) return c + 0x28
 
```
10400; C; 10428; # DESERET CAPITAL LETTER LONG I
10401; C; 10429; # DESERET CAPITAL LETTER LONG E
10402; C; 1042A; # DESERET CAPITAL LETTER LONG A
10403; C; 1042B; # DESERET CAPITAL LETTER LONG AH
10404; C; 1042C; # DESERET CAPITAL LETTER LONG O
10405; C; 1042D; # DESERET CAPITAL LETTER LONG OO
10406; C; 1042E; # DESERET CAPITAL LETTER SHORT I
10407; C; 1042F; # DESERET CAPITAL LETTER SHORT E
10408; C; 10430; # DESERET CAPITAL LETTER SHORT A
10409; C; 10431; # DESERET CAPITAL LETTER SHORT AH
1040A; C; 10432; # DESERET CAPITAL LETTER SHORT O
1040B; C; 10433; # DESERET CAPITAL LETTER SHORT OO
1040C; C; 10434; # DESERET CAPITAL LETTER AY
1040D; C; 10435; # DESERET CAPITAL LETTER OW
1040E; C; 10436; # DESERET CAPITAL LETTER WU
1040F; C; 10437; # DESERET CAPITAL LETTER YEE
10410; C; 10438; # DESERET CAPITAL LETTER H
10411; C; 10439; # DESERET CAPITAL LETTER PEE
10412; C; 1043A; # DESERET CAPITAL LETTER BEE
10413; C; 1043B; # DESERET CAPITAL LETTER TEE
10414; C; 1043C; # DESERET CAPITAL LETTER DEE
10415; C; 1043D; # DESERET CAPITAL LETTER CHEE
10416; C; 1043E; # DESERET CAPITAL LETTER JEE
10417; C; 1043F; # DESERET CAPITAL LETTER KAY
10418; C; 10440; # DESERET CAPITAL LETTER GAY
10419; C; 10441; # DESERET CAPITAL LETTER EF
1041A; C; 10442; # DESERET CAPITAL LETTER VEE
1041B; C; 10443; # DESERET CAPITAL LETTER ETH
1041C; C; 10444; # DESERET CAPITAL LETTER THEE
1041D; C; 10445; # DESERET CAPITAL LETTER ES
1041E; C; 10446; # DESERET CAPITAL LETTER ZEE
1041F; C; 10447; # DESERET CAPITAL LETTER ESH
10420; C; 10448; # DESERET CAPITAL LETTER ZHEE
10421; C; 10449; # DESERET CAPITAL LETTER ER
10422; C; 1044A; # DESERET CAPITAL LETTER EL
10423; C; 1044B; # DESERET CAPITAL LETTER EM
10424; C; 1044C; # DESERET CAPITAL LETTER EN
10425; C; 1044D; # DESERET CAPITAL LETTER ENG
10426; C; 1044E; # DESERET CAPITAL LETTER OI
10427; C; 1044F; # DESERET CAPITAL LETTER EW
```

**Comment by iSazonov** [comment]
Source is two surrogates (high and low). So we have to calculate codepoint, then +0x28, then decode to new surrogate pair. It looks very-very slow . Or is the case very rare and we could be slow?
Also what about codes below 10427?

**Comment by tarekgh** [comment]
The check is very simple like 

```C#

if (ch == 0xD801 && (cl >= 0xDC00 && cl <= 0xDC27 ))
   return 0xD801, cl+0x28;
```

**Comment by iSazonov** [comment]
I created https://github.com/iSazonov/SCF/ with my current code.
Three mapping is used:
- for chars <= 0x5ff - 1-level mapping (~3 Kb)
- for chars > 0x5ff (not surrogates) - 2-level mapping (~10 Kb)
- for surrogates - 2-level mapping (~6 Kb)
Total size ~19 Kbyte.

Status: 
- surrogates is not still processed.
- working on optimization of comparer CompareUsingSimpleCaseFolding() method.

Best performance results (2nd test uses Russian chars):
``` ini

BenchmarkDotNet=v0.11.3, OS=Windows 10.0.17763.134 (1809/October2018Update/Redstone5)
Intel Core i5-2410M CPU 2.30GHz (Sandy Bridge), 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=2.1.403
  [Host]    : .NET Core 2.1.5 (CoreCLR 4.6.26919.02, CoreFX 4.6.26919.02), 64bit RyuJIT
  RyuJitX64 : .NET Core 2.1.5 (CoreCLR 4.6.26919.02, CoreFX 4.6.26919.02), 64bit RyuJIT

Job=RyuJitX64  Jit=RyuJit  Platform=X64  

```
|                Method |         StrA |         StrB |     Mean |     Error |    StdDev | Ratio |
|---------------------- |------------- |------------- |---------:|----------:|----------:|------:|
|         **CoreFXCompare** | **CaseFolding1** | **cASEfOLDING2** | **42.09 ns** | **0.3242 ns** | **0.2874 ns** |  **1.00** |
| SimpleCaseFoldCompare | CaseFolding1 | cASEfOLDING2 | 37.89 ns | 0.2617 ns | 0.2448 ns |  0.90 |
|                       |              |              |          |           |           |       |
|         **CoreFXCompare** | **ЯяЯяЯяЯяЯяЯ1** | **яЯяЯяЯяЯяЯя2** | **89.44 ns** | **0.7416 ns** | **0.6574 ns** |  **1.00** |
| SimpleCaseFoldCompare | ЯяЯяЯяЯяЯяЯ1 | яЯяЯяЯяЯяЯя2 | 37.68 ns | 0.2436 ns | 0.2160 ns |  0.42 |

It is faster than CoreFX OrdinalIgnoreCase compare.
Perhaps @stephentoub will be interested in the performance result.

A problem where I need help from MSFT experts: with any changes in [CompareUsingSimpleCaseFolding() method](https://github.com/iSazonov/SCF/blob/598366ee3a1ed2c0e724460f332d43539288a6bd/src/SimpleCaseFolding.cs#L144) performance degrades. I can not found how add processing surrogates without lost performance. As soon as I add [this call](https://github.com/iSazonov/SCF/blob/598366ee3a1ed2c0e724460f332d43539288a6bd/src/SimpleCaseFolding.cs#L203), performance drops even though this code is not executed at all (there are no surrogates in the test line!). I tried different ways (goto-s, inlining ...) but failed :-(
Perhaps @AndyAyersMS could help or ping right experts.

**Comment by iSazonov** [comment]
I see .Net Core team updated Marvin https://source.dot.net/#System.Private.CoreLib/shared/System/Marvin.OrdinalIgnoreCase.cs,40
to increase performance for OrdinalIgnoreCase if chars is ASCII.
With the simple case folding the Marvin code could be faster for chars <= 0x5ff that is most of the commonly used characters.

**Comment by iSazonov** [comment]
Benchmarks for simple case folding vs String.ToUpperInvariant
``` ini

BenchmarkDotNet=v0.11.3, OS=Windows 10.0.17763.253 (1809/October2018Update/Redstone5)
Intel Core i3-4150 CPU 3.50GHz (Haswell), 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=2.1.403
  [Host]    : .NET Core 2.1.5 (CoreCLR 4.6.26919.02, CoreFX 4.6.26919.02), 64bit RyuJIT
  RyuJitX64 : .NET Core 2.1.5 (CoreCLR 4.6.26919.02, CoreFX 4.6.26919.02), 64bit RyuJIT

Job=RyuJitX64  Jit=RyuJit  Platform=X64  

```
|                 Method |         StrA |     Mean |     Error |    StdDev | Ratio | RatioSD |
|----------------------- |------------- |---------:|----------:|----------:|------:|--------:|
| **CoreFXToUpperInvariant** | **CaseFolding1** | **40.21 ns** | **0.2205 ns** | **0.1722 ns** |  **1.00** |    **0.00** |
| CoreFXToLowerInvariant | CaseFolding1 | 34.56 ns | 0.1502 ns | 0.1254 ns |  0.86 |    0.00 |
|         SimpleCaseFold | CaseFolding1 | 29.54 ns | 0.1277 ns | 0.1066 ns |  0.73 |    0.00 |
|                        |              |          |           |           |       |         |
| **CoreFXToUpperInvariant** | **ЯяЯяЯяЯяЯяЯ1** | **80.33 ns** | **1.3473 ns** | **1.1943 ns** |  **1.00** |    **0.00** |
| CoreFXToLowerInvariant | ЯяЯяЯяЯяЯяЯ1 | 81.19 ns | 1.7442 ns | 2.0086 ns |  1.01 |    0.03 |
|         SimpleCaseFold | ЯяЯяЯяЯяЯяЯ1 | 29.48 ns | 0.1668 ns | 0.1393 ns |  0.37 |    0.01 |

**Comment by iSazonov** [comment]
I grub Marvin to my local (with minor changes to get compile) and replace ToUpperInvariant() with simple case folding (I haven't done any more optimizations.).  We can see again great performance win for chars > 0x5ff.
``` ini

BenchmarkDotNet=v0.11.3, OS=Windows 10.0.17763.253 (1809/October2018Update/Redstone5)
Intel Core i3-4150 CPU 3.50GHz (Haswell), 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=3.0.100-preview-010076
  [Host]    : .NET Core 3.0.0-preview-27310-6 (CoreCLR 4.6.27220.01, CoreFX 4.7.19.6004), 64bit RyuJIT
  RyuJitX64 : .NET Core 3.0.0-preview-27310-6 (CoreCLR 4.6.27220.01, CoreFX 4.7.19.6004), 64bit RyuJIT

Job=RyuJitX64  Jit=RyuJit  Platform=X64  

```
|                        Method |         StrA |     Mean |     Error |    StdDev | Ratio |
|------------------------------ |------------- |---------:|----------:|----------:|------:|
| **CoreFXMarvinOrdinalIgnoreCase** | **CaseFolding1** | **12.18 ns** | **0.0927 ns** | **0.0724 ns** |  **1.00** |
|    SCFMarvinOrdinalIgnoreCase | CaseFolding1 | 11.28 ns | 0.1656 ns | 0.1468 ns |  0.93 |
|                               |              |          |           |           |       |
| **CoreFXMarvinOrdinalIgnoreCase** | **ЯяЯяЯяЯяЯяЯ1** | **96.34 ns** | **0.9374 ns** | **0.7319 ns** |  **1.00** |
|    SCFMarvinOrdinalIgnoreCase | ЯяЯяЯяЯяЯяЯ1 | 39.45 ns | 0.5140 ns | 0.4557 ns |  0.41 |

**Comment by iSazonov** [comment]
I moved the project from .Net Core 2.1 to 3.0-preview (latest daily build) and overall performance noticeably higher for all perf tests!

**Comment by iSazonov** [comment]
@ahsonkhan @tarekgh what can I do next?

**Comment by tarekgh** [comment]
As I originally suggested to have this in NuGet package and let's see the demand on this library, and if we see enough demand we can consider moving the functionality to the core. My concern to move this to core from now is the data size and not clear if this feature will be used by many people.

**Comment by iSazonov** [comment]
> As I originally suggested to have this in NuGet package

I am ready. Is there a manual?
Package name and namespace is a question. Need help.

>  My concern to move this to core from now is the data size 

Yes, data size is relatively big. On the other hand, benefits are also significant - this code may be even faster than OrdinalIgnoreCase for char < 0x7f. It is even possible that MSFT experts can make it even faster.

> and not clear if this feature will be used by many people.

I don't think that the package will be used many people directly because most of benefits that we can get is in CoreCLR/CoreFX - I listed them earlier and I guess there are even more.

I'd ask MSFT team to look the approach in depth. I believe many APIs can get performance benefits (over 2x).

**Comment by tarekgh** [comment]
I recall the libraries in corefxlab got packaged as nuget package. @ahsonkhan may confirm that. 

@joshfree can advice if we can host this library in corefxlab. but in general, you may look at https://www.nuget.org/ for the documentation how you can create your own NuGet library.

**Comment by ahsonkhan** [comment]
Libraries from corefxlab are built and published to MyGet: https://dotnet.myget.org/gallery/dotnet-corefxlab (as long as the project is included in the corefxlab.sln solution). Consider following an existing project as a starting point to make it easier to build the package and publish it.

Feel free to ping me on the PR/branch that attempts to add the new project to this repo and I can provide feedback.

> Package name and namespace is a question

Usually our package name == project name. For example: https://github.com/dotnet/corefxlab/blob/master/src/System.Text.Primitives/System.Text.Primitives.csproj

For the name itself, I would defer to what @tarekgh thinks. Maybe we can add the APIs to the existing System.Text.Primitives package, if it makes sense.

For the namespace, it looks like `System.Globalization` could work, but I am not sure.

What namespace do the related APIs/types within corefx live in? We should use something similar.

**Comment by tarekgh** [comment]
I would suggest assembly and namespace name like System.Globalization.Utils (or something similar). and if this going one day to move to core, we'll include it in System.Globalization.Extensions assembly (or maybe in coreclr if proven need some core dependencies somehow).

**Comment by iSazonov** [comment]
@ahsonkhan @tarekgh Thanks for clarify! Will do.

@tarekgh Could you please comment System.Text vs System.Globalization.Utils (or something similar)
If you prefer System.Globalization can it be System.Globalization.Fold or it is very special?

**Comment by tarekgh** [comment]
> Could you please comment System.Text/System.Text.Fold vs System.Globalization.Utils (or something similar)

Folding is in the boarder line between Text and Globalization. If your library regarding folding only, I would say this can go to System.Text as I am seeing this functionality like casing which we have it in System.Text. if you are going to provide more functionality like Comparing strings (using folding), that may go to globalization. I don't mind to have it in System.Text though if the name of the methods reflect the folding. 

in short, I don't mind adding that to System.Text.Primitives too. Actually thinking more I would prefer System.Text.Primitives more. Just make sure you use the System.Text.Primitives as a namespace too.

> As for moving to CoreFX/CoreCRL here is folding, comparer and span APIs which are likely to be placed in the appropriate namespaces and assemblies.

Right. when getting stuff in core we place it under the right namespace.

**Comment by iSazonov** [comment]
Thanks! I'll use System.Text.Primitives if it is not for UTF-8 only.

**Comment by iSazonov** [comment]
@tarekgh System.Text.Primitives uses System.Buffers.Text namespace.
We could use System.Text.Folding for start. Can we still put it in System.Text.Primitives or better create new package?

**Comment by joshfree** [comment]
> @joshfree can advice if we can host this library in corefxlab.

@tarekgh The bar for corefxlab prototyping is that we believe the prototype may be productized and promoted to dotnet/corefx (or similar) shipping repos in the future.

**Comment by tarekgh** [comment]
> We could use System.Text.Folding for start. Can we still put it in System.Text.Primitives or better create new package?

Maybe it is better idea to have System.Text.CaseFolding library and use the same name for the namespace too. I looked more at corefxlab and we have multiple System.Text libraries for different functionality so it will make sense to introduce System.Text.CaseFolding

**Comment by iSazonov** [comment]
@tarekgh @ahsonkhan PR was pushed.

**Comment by iSazonov** [comment]
@tarekgh Could you please share an information what is InvariantCulture mapping tables? I don't found any information in Unicode standard. Seems it is internal Windows implementation like different "C" locales on different Unix-es. In the case Simple Case Folding looks good standardized candidate for unified implementation of invariant comparisons for all platforms.

**Comment by tarekgh** [comment]
> Could you please share an information what is InvariantCulture mapping tables?

I am not sure what exactly the ask here. are you asking about the casing mapping for invariant? we don't carry any data and we just use the underlying OS for doing the casing.

**Comment by iSazonov** [comment]
>  you asking about the casing mapping for invariant?

Yes, my ask about underlying Windows implementation.

**Comment by tarekgh** [comment]
> Yes, my ask about underlying Windows implementation.

Windows has the casing tables file. it is easy to generate the mapping through looping the whole Unicode range and do ToUpper, ToLower using Invariant culture. by the way, we are talking here about casing and not case folding. just in case we confuse both together.

**Comment by iSazonov** [comment]
Sorry for the inaccuracy of my question. My question come from https://github.com/PowerShell/PowerShell/issues/8064 and https://github.com/dotnet/corefx/issues/28611 - 
i.e. comparison. Using InvariantCulture gives different results because of differences in underlying OS implementations. So I think of Simple Case Folding as a standardized alternative.

**Comment by tarekgh** [comment]
> Using InvariantCulture gives different results because of differences in underlying OS implementations. So I think of Simple Case Folding as a standardized alternative.

case folding is not going to help in the linguistic (non-ordinal) operations.

**Comment by iSazonov** [comment]
@ahsonkhan Can we publish the package?

We measure performance with latest 3.0 Preview6 and there is no performance benefits or regressions.
I believe there can still be made performance improvements but only by experts with deeper skills than I have.

**Comment by ahsonkhan** [comment]
> Can we publish the package?

Its already available on the corefxlab MyGet feed: https://dotnet.myget.org/feed/dotnet-corefxlab/package/nuget/System.Text.CaseFolding/0.1.1-e190517-1
